### ISSUE TITLE
High memory usage while compiling with VS2015

### ISSUE THREAD

**Issue by roigcarlo** [ddemidov/amgcl#43]
Dear Denis,

Recently we have been having some problems while compiling AMGCL with Kratos using Visual studio 2015 as it complains that it runs out of memory. We have managed to bypass it by commenting https://github.com/ddemidov/amgcl/blob/master/amgcl/util.hpp#L126.

So our current code looks like:

```C++
struct empty_params {
    empty_params() {}
    empty_params(const boost::property_tree::ptree &p) {
        // AMGCL_PARAMS_CHECK(p, BOOST_PP_EMPTY());
    }
    void get(boost::property_tree::ptree&, const std::string&) const {}
};
```
But @RiccardoRossi found that this was actually added as a fix for visual studio in https://github.com/ddemidov/amgcl/commit/040b9f58067a9d6fd79ed4981aeb3aac3692fd9a.

We are not sure of what this does or if it is safe to comment at all. Any clue on the topic?

**Comment by ddemidov** [comment]
Hi Carlos,

Can you please see if ff5f865 fixes this for you?

**Comment by ddemidov** [comment]
Riccardo commented here:

> ￼hi Carlos, Denis,
actually i was compiling this afternoon on MiguelAngel's coputer, and it fails even after commenting this.
the latest error is saw is: 9>c:\users\ignasi\kratos\external_libraries\amgcl\relaxation\chebyshev.hpp(91): fatal error C1002: el compilador no tiene espacio en el montón en el paso 2
i am now starting to suspect that the problem is kratos+amgcl occupying a lot of memory, not just AMGCL :-(
i will try to have amgcl compilation alone on a single cpp, let's see if this helps

And my reply:

> That may actually be the case with the combined amount of templated code kratos and amgcl have together.

**Comment by ddemidov** [comment]
Could this be relevant? http://stackoverflow.com/a/33469122/996848

**Comment by ddemidov** [comment]
@RiccardoRossi reported that he managed to solve the memory issue by moving amgcl-related code into a separate code unit, but I went ahead and merged 0d4edd2 anyway, since that solves a couple of warnings in Visual C++.
